control/noise/handshake.go

```diff
@@ -0,0 +1,430 @@
+		handshakeHash: s.h,
+		tx: txState{
+			cipher: c2,
+		},
+		rx: rxState{
+			cipher: c1,
+		},
+	}, nil
+}
+
+// symmetricState is the SymmetricState object from the Noise protocol
+// spec. It contains all the symmetric cipher state of an in-flight
+// handshake. Field names match the variable names in the spec.
+type symmetricState struct {
+	h  [blake2s.Size]byte
+	ck [blake2s.Size]byte
+
+	k [chp.KeySize]byte
+	n uint64
+
+	mixer hash.Hash // for updating h
+}
+
+// Initialize sets s to the initial handshake state, prior to
+// processing any Noise messages.
+func (s *symmetricState) Initialize() {
+	if s.mixer != nil {
+		panic("symmetricState cannot be reused")
+	}
+	s.h = blake2s.Sum256([]byte(protocolName))
+	s.ck = s.h
+	s.k = [chp.KeySize]byte{}
+	s.n = invalidNonce
+	s.mixer = newBLAKE2s()
+}
+
+// MixHash updates s.h to be BLAKE2s(s.h || data), where || is
+// concatenation.
+func (s *symmetricState) MixHash(data []byte) {
+	s.mixer.Reset()
+	s.mixer.Write(s.h[:])
+	s.mixer.Write(data)
+	s.mixer.Sum(s.h[:0]) // TODO: check this actually updates s.h correctly...
+}
+
+// MixDH updates s.ck and s.k with the result of X25519(priv, pub).
+//
+// MixDH corresponds to MixKey(X25519(...))) in the spec. Implementing
+// it as a single function allows for strongly-typed arguments that
+// reduce the risk of error in the caller (e.g. invoking X25519 with
+// two private keys, or two public keys), and thus producing the wrong
+// calculation.
+func (s *symmetricState) MixDH(priv key.Private, pub key.Public) error {
+	// TODO(danderson): check that this operation is correct. The docs
+	// for X25519 say that the 2nd arg must be either Basepoint or the
+	// output of another X25519 call.
+	//
+	// I think this is correct, because pub is the result of a
+	// ScalarBaseMult on the private key, and our private key
+	// generation code clamps keys to avoid low order points. I
+	// believe that makes pub equivalent to the output of
+	// X25519(privateKey, Basepoint), and so the contract is
+	// respected.
+	keyData, err := curve25519.X25519(priv[:], pub[:])
+	if err != nil {
+		return fmt.Errorf("computing X25519: %w", err)
+	}
+
+	r := hkdf.New(newBLAKE2s, keyData, s.ck[:], nil)
+	if _, err := io.ReadFull(r, s.ck[:]); err != nil {
+		return fmt.Errorf("extracting ck: %w", err)
+	}
+	if _, err := io.ReadFull(r, s.k[:]); err != nil {
+		return fmt.Errorf("extracting k: %w", err)
+	}
+	s.n = 0
+	return nil
+}
+
+// EncryptAndHash encrypts plaintext into ciphertext (which must be
+// the correct size to hold the encrypted plaintext) using the current
+// s.k, mixes the ciphertext into s.h, and returns the ciphertext.
+func (s *symmetricState) EncryptAndHash(ciphertext, plaintext []byte) {
+	if s.n == invalidNonce {
+		// Noise in general permits writing "ciphertext" without a
+		// key, but in IK it cannot happen.
+		panic("attempted encryption with uninitialized key")
+	}
+	if len(ciphertext) != len(plaintext)+poly1305.TagSize {
```

Comment:
```suggestion
	if len(ciphertext) != len(plaintext)+chp.Overhead {
```

and elsewhere, the poly1305 package is deprecated.